File: #0739.daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {

        int size = temperatures.size();
        vector<int> result(size);
        stack<int> stk;
        for (int i = size -1; i >= 0; i--) {
            while (!stk.empty() &&
                  temperatures[stk.top()] <= temperatures[i]) {
                stk.pop();
            }
            result[i] = stk.empty() ? 0 : stk.top() - i;
            stk.push(i);
        }
        return result;
    }
};

File: #0739.daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    let stack = [];
    for (let i = 0; i < temperatures.length; i++) {
        while (temperatures[i] > temperatures[stack[0]]) {
            let shifted = stack.shift();
            temperatures[shifted] = i - shifted;
        }
        stack.unshift(i);
    }

    while (stack.length) {
        temperatures[stack.shift()] = 0;
    }
    return temperatures;
};
